Common writer for plink and ICF

benjeffery opened the pull request:
This WIP PR attempts to use a single schema and writer class for both ICF and plink encoding. The tests are almost completely unchanged. 

This needs some polish - for example I've only adapted the `Writer.encode_X` methods needed for plink. They should be consistent and ICF specific stuff moved out of the common writer.

Plink conversion is also missing the indexing.

jeromekelleher commented on bio2zarr/vcf2zarr/vcz.py:
```diff
@@ -240,993 +80,156 @@ def convert_local_allele_field_types(fields):
-
-        if local_alleles:
-            array_specs = convert_local_allele_field_types(array_specs)
+def generate_schema(
```
Shouldn't this just be a method of the ICF/Plink formats? The vcz module shouldn't be creating schemas I think.

jeromekelleher commented on bio2zarr/vcf2zarr/vcz.py:
```diff
@@ -240,993 +80,156 @@ def convert_local_allele_field_types(fields):
-            j = gt_phased.next_buffer_row()
-            icf.sanitise_value_int_1d(
-                gt_phased.buff, j, value[:, -1] if value is not None else None
+    array_specs = [
```
This definitely seems like it should be a method of the ICF

jeromekelleher commented on bio2zarr/vcf2zarr/vcz.py:
```diff
@@ -1248,7 +251,7 @@ def mkschema(
 
 
 def encode(
```
These high-level methods are breaking the clarity of the division of labour here; it is probably too much to move them out in this PR, but we should look at moving them somewhere else. The current modules should just be concerned with writing VCZ efficiently, and not know any thing about where that data comes from.

jeromekelleher commented on bio2zarr/writer.py:
```diff
@@ -0,0 +1,813 @@
+import dataclasses
```
I'm not sure I see the value of this module - why not keep the VCZ writing logic in vcz.py (which we could move to root of the package hierarchy)?

jeromekelleher commented on tests/test_vcf_examples.py:
```diff
@@ -109,6 +109,8 @@ def test_float_info_fields(self, ds):
             dtype=np.float32,
         )
         values = ds["variant_AF"].values
+        print(values)
```
stray print

jeromekelleher reviewed (commented):
Looks great! I'm not sure I follow the divison of labour, though.

How about the following structure:

- bio2zarr/vcz.py. This contains the definition of a VcfZarrSchema and the VcfZarrWriter class. It does not know about icf, plink or any other format.


ICF anda PlinkFormat then have a ``generate_schema`` method. So, icf.py and plink.py will both need to import vcz.py, but that makes sense as that is the format that they are targeting.

Does that makes sense?

benjeffery commented:
Yes, I meant to have a re-org and forgot to.
I've removed make the schema generators member functions of their sources. I've removed `vcz.py` with the `encode_*` methods going into `icf.py`.
I'd like to keep `schema.py` and `writer.py` separate - it is most of what was in `vcz.py` but now at the top-level.

Will double check the plink details next.

jeromekelleher commented on bio2zarr/writer.py:
```diff
@@ -615,13 +160,14 @@ class VcfZarrWriteSummary(core.JsonDataclass):
 
 class VcfZarrWriter:
-    def __init__(self, path):
+    def __init__(self, source_type, path):
```
Why does the writer need to know the class of the source ahead of init?

benjeffery replied:
`load_metadata` needs it, which can be called by e.g. `encode_parition` without `init` being called.

benjeffery commented on bio2zarr/plink.py:
```diff
@@ -1,177 +1,163 @@
+            yield alleles
+
+    def iter_field(self, field_name, shape, start, stop):
+        data = {
```
Yes, when I wrote this I thought there might be other fields that could be added.

jeromekelleher replied:
Let's switch to the obvious version so - it took me a good 30 seconds to figure out what was happening here.

benjeffery commented on bio2zarr/plink.py:
```diff
@@ -1,177 +1,163 @@
+        m = self.bed.sid_count
+        logging.info(f"Scanned plink with {n} samples and {m} variants")
+
+        # FIXME
```
Fixed in 66a86fc

benjeffery commented on bio2zarr/vcf2zarr/icf.py:
```diff
@@ -8,12 +8,16 @@
-from .. import constants, core, provenance, vcf_utils
+from bio2zarr import schema
+
+from .. import constants, core, provenance, vcf_utils, writer
```
Fixed in 66a86fc

benjeffery commented on bio2zarr/vcf2zarr/icf.py:
```diff
@@ -916,6 +956,229 @@ def num_samples(self):
+    def generate_schema(
+        self, variants_chunk_size=None, samples_chunk_size=None, local_alleles=None
+    ):
+        # Import schema here to avoid circular import
```
Fixed in 66a86fc

benjeffery commented on bio2zarr/vcf2zarr/icf.py:
```diff
@@ -916,6 +956,229 @@ def num_samples(self):
+
+        m = self.num_records
+        n = self.num_samples
+        if samples_chunk_size is None:
```
Fixed in 66a86fc

benjeffery commented:
You're right, I was missing that point!
I've combined the writer and schema into `vcz.py` and removed the `vcf2zarr` module.

jeromekelleher commented:
This looks great, a really big step forward in terms of clarity and extensibility. 

I think it's ready to merge, but I'm still uneasy about flipping the A1/A2 alleles in plink. Unless there's a good reason (beyond it being less confusing) I think we should keep it the way it is, as the rules are quite complicated and there could easily be untested corner cases that the new behaviour differs on. I took the original code from sgkit, and I think it was pretty well used there, so it should be reasonably correct.

benjeffery commented on bio2zarr/plink.py:
```diff
@@ -1,69 +1,60 @@
+        self.root_attrs = {}
+
+    def iter_alleles(self, start, stop, num_alleles):
+        ref_field = self.bed.allele_1
```
It seems there is no fixed convention and we'll have to allow this to be user configurable?
From reading https://www.cog-genomics.org/plink/1.9/data it seems by default, PLINK often assigns the major allele as A2 and the minor allele as A1, but this can be overridden with flags like --keep-allele-order, --real-ref-alleles, --a1-allele, or --a2-allele.

jeromekelleher reviewed (approved):
Looks good to me! I'm still a bit iffy about changing the plink code, but I guess if the tests are passing it's probably OK. Ready for a squash and merge I think.

jeromekelleher commented:
Hmm, looks like something funny is happening with numcodecs exact sizes again. Did we just get a release or something?

benjeffery commented:
> Hmm, looks like something funny is happening with numcodecs exact sizes again. Did we just get a release or something?

Yep about an hour ago. I've filed #347
